File: backend/app/schemas/ai_quiz.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class AIQuizQuestion(BaseModel):
    question_index: int
    question_text: str
    answer_type: str
    options: list[str] | None = None


class AIQuizGenerateRequest(BaseModel):
    quiz_type: str = "stress"
    language: str = "en"


class AIQuizSessionResponse(BaseModel):
    id: int
    quiz_type: str
    status: str
    generated_questions: list[AIQuizQuestion]
    created_at: datetime
    updated_at: datetime


class AIQuizAnswerItem(BaseModel):
    question_index: int
    question_text: str
    answer_text: str
    score: int | None = Field(default=None, ge=0, le=4)


class AIQuizSubmitRequest(BaseModel):
    answers: list[AIQuizAnswerItem]
    language: str = "en"


class AIQuizMicroPractice(BaseModel):
    title: str
    description: str
    estimated_time: str
    action: Literal["journal", "mood", "goals"] | None = None


class AIQuizActionPlan(BaseModel):
    quiz_type: str
    result_level: str
    steps: list[str]
    micro_practices: list[AIQuizMicroPractice]
    reflection_prompt: str
    suggested_goal: str | None = None
    supportive_message: str


class AIQuizTrend(BaseModel):
    trend_direction: Literal["improved", "worsened", "stable", "first_time"]
    previous_score: float | None = None
    current_score: float
    score_difference: float | None = None
    explanation: str


class AIQuizResultResponse(BaseModel):
    id: int
    session_id: int
    quiz_type: str | None = None
    overall_score: float
    severity_level: str
    insight: str
    recommendation: str
    practice: str
    recommendations: list[str] | None = None
    micro_practices: list[AIQuizMicroPractice] | None = None
    action_plan: AIQuizActionPlan | None = None
    trend_direction: str | None = None
    previous_score: float | None = None
    score_difference: float | None = None
    trend_explanation: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AIQuizDetailResponse(BaseModel):
    session: AIQuizSessionResponse
    result: AIQuizResultResponse | None = None


class AIQuizTypeResponse(BaseModel):
    key: str
    title: str
    description: str
    estimated_minutes: int
    emoji: str | None = None
    status: Literal["not_started", "completed", "completed_recently"]
    latest_result_id: int | None = None
    latest_score: float | None = None
    latest_level: str | None = None
    latest_completed_at: datetime | None = None


class AIQuizHistoryItem(BaseModel):
    result_id: int
    session_id: int
    quiz_type: str
    quiz_title: str
    completed_at: datetime
    score: float
    severity_level: str
    summary: str
    has_recommendations: bool
    has_action_plan: bool
    trend_direction: str | None = None
    previous_score: float | None = None
    score_difference: float | None = None


class AIQuizLatestActionPlanResponse(BaseModel):
    result_id: int
    session_id: int
    quiz_type: str
    quiz_title: str
    created_at: datetime
    score: float
    severity_level: str
    summary: str
    next_actions: list[str]
    action_plan: AIQuizActionPlan

File: mobile/src/api/aiQuiz.ts
import { apiFetch } from './client';

export type AiQuizQuestion = {
  question_index: number;
  question_text: string;
  answer_type: string;
  options?: string[] | null;
};

export type AiQuizGenerateResponse = {
  id: number;
  quiz_type: string;
  status: string;
  generated_questions: AiQuizQuestion[];
  created_at: string;
  updated_at: string;
};

export type AiQuizMicroPractice = {
  title: string;
  description: string;
  estimated_time: string;
  action?: 'journal' | 'mood' | 'goals' | null;
};

export type AiQuizActionPlan = {
  quiz_type: string;
  result_level: string;
  steps: string[];
  micro_practices: AiQuizMicroPractice[];
  reflection_prompt: string;
  suggested_goal?: string | null;
  supportive_message: string;
};

export type AiQuizResultDetail = {
  id: number;
  session_id: number;
  quiz_type?: string | null;
  overall_score: number;
  severity_level: string;
  insight: string;
  recommendation: string;
  practice: string;
  recommendations?: string[] | null;
  micro_practices?: AiQuizMicroPractice[] | null;
  action_plan?: AiQuizActionPlan | null;
  trend_direction?: string | null;
  previous_score?: number | null;
  score_difference?: number | null;
  trend_explanation?: string | null;
  created_at: string;
  updated_at: string;
};

export type AiQuizDetailResponse = {
  session: AiQuizGenerateResponse;
  result: AiQuizResultDetail | null;
};

export type AiQuizSubmitAnswer = {
  question_index: number;
  question_text: string;
  answer_text: string;
  score?: number | null;
};

export type AiQuizType = {
  key: string;
  title: string;
  description: string;
  estimated_minutes: number;
  emoji?: string | null;
  status: 'not_started' | 'completed' | 'completed_recently';
  latest_result_id?: number | null;
  latest_score?: number | null;
  latest_level?: string | null;
  latest_completed_at?: string | null;
};

export type AiQuizHistoryItem = {
  result_id: number;
  session_id: number;
  quiz_type: string;
  quiz_title: string;
  completed_at: string;
  score: number;
  severity_level: string;
  summary: string;
  has_recommendations: boolean;
  has_action_plan: boolean;
  trend_direction?: string | null;
  previous_score?: number | null;
  score_difference?: number | null;
};

export type AiQuizLatestActionPlan = {
  result_id: number;
  session_id: number;
  quiz_type: string;
  quiz_title: string;
  created_at: string;
  score: number;
  severity_level: string;
  summary: string;
  next_actions: string[];
  action_plan: AiQuizActionPlan;
};

export type AiQuizResult = AiQuizResultDetail | null;

export function getAiQuizTypes() {
  return apiFetch<AiQuizType[]>('/ai-quiz/types', { method: 'GET', auth: true });
}

export function generateAiQuiz(quiz_type = 'stress', language = 'en') {
  return apiFetch<AiQuizGenerateResponse>('/ai-quiz/generate', {
    method: 'POST',
    auth: true,
    body: JSON.stringify({ quiz_type, language }),
  });
}

export function getAiQuizSession(session_id: number) {
  return apiFetch<AiQuizDetailResponse>(`/ai-quiz/${session_id}`, {
    method: 'GET',
    auth: true,
  });
}

export function submitAiQuiz(session_id: number, answers: AiQuizSubmitAnswer[], language = 'en') {
  return apiFetch<AiQuizResultDetail>(`/ai-quiz/${session_id}/submit`, {
    method: 'POST',
    auth: true,
    body: JSON.stringify({ answers, language }),
  });
}

export function getAiQuizHistory(limit = 25) {
  return apiFetch<AiQuizHistoryItem[]>(`/ai-quiz/history?limit=${limit}`, {
    method: 'GET',
    auth: true,
  });
}

export function getAiQuizResult(resultId: number) {
  return apiFetch<AiQuizResultDetail>(`/ai-quiz/history/${resultId}`, {
    method: 'GET',
    auth: true,
  });
}

export function getLatestAiQuizActionPlan() {
  return apiFetch<AiQuizLatestActionPlan | null>('/ai-quiz/latest-action-plan', {
    method: 'GET',
    auth: true,
  });
}

export function saveAiQuizActionPlan(resultId: number) {
  return apiFetch<AiQuizLatestActionPlan>(`/ai-quiz/results/${resultId}/save-action-plan`, {
    method: 'POST',
    auth: true,
    body: JSON.stringify({}),
  });
}
